### PR TITLE
Use Parity Treasury & Tips

### PR DESCRIPTION
**Pull request by AurevoirXavier**
### Changes
- Update `transaction_version`
- Remove Old `Bounties` and Add New `Bounties`
- Migrate `DarwiniaTreasury` to `Treasury`, `Instance2Treasury` and `Instance1Tips`
	- **Darwinia**
		- `ProposalCount` Old Data
			***Empty***
		- `Proposals` Old Data
			***Empty***
		- `Approvals` Old Data
			***Empty***
		- `Tips` Old Data
			```json
			[
				[
					"0x327473e57e0ba6cc39e52adc79597a772c5de123c468aef7f3ac2ab3a76f87ce065ece05537dcc41c10b609b7793d4d11e32a2e31d11f8eacb0784dc35cc91e4dc7c364ce53f573c",
					"0xee6d208478391c285c1759f7ad54f6ade35f3835d059053f7527a3924b65411c501112f2f16ba59caf810432abbc8f1b75599675b5e07f26cac60fa5b50f8b42501112f2f16ba59caf810432abbc8f1b75599675b5e07f26cac60fa5b50f8b420090d003000000000000000000000000000001"
				],
				[
					"0x327473e57e0ba6cc39e52adc79597a772c5de123c468aef7f3ac2ab3a76f87ce099298a59dca0de1bd185f9aac4bd316ac126511906ed3cb97f83edcb45942187ed554b48d14c2cb",
					"0xc47094c8866909e63651aea39a57844eb4262d293796ab651bd5d64f7201d7cd501112f2f16ba59caf810432abbc8f1b75599675b5e07f26cac60fa5b50f8b42501112f2f16ba59caf810432abbc8f1b75599675b5e07f26cac60fa5b50f8b4240c06503000000000000000000000000000001"
				],
				[
					"0x327473e57e0ba6cc39e52adc79597a772c5de123c468aef7f3ac2ab3a76f87ce2fce5d121b18145bd42ad06ad670ad72a5ab2275fb12a414bb89d41ac78d5bce76c822ea92e7437d",
					"0x0095e6fbc1a9e1c46b467c20f936afd09eaddec481acf7a7af9fb0568f1ede3d929b1ca32c496c4a85e9a90d2b5cd4b37f236597bca732f03b561f2855e8df0b929b1ca32c496c4a85e9a90d2b5cd4b37f236597bca732f03b561f2855e8df0b804a5d05000000000000000000000000000001"
				],
				[
					"0x327473e57e0ba6cc39e52adc79597a772c5de123c468aef7f3ac2ab3a76f87ce76c6adafe24590724cb1c3401c79aa9e629e532472adaad11ac757276c9b3dd9064daeb46e728a64",
					"0x2b1cf536e9043fc21205e9b698bacde8c286e2e14c4dfb92fe4bcd0de74ecabaa60ff79f17777fe50fae99e873a56596da8fa1c3d65331600e211ebac737cf7fa60ff79f17777fe50fae99e873a56596da8fa1c3d65331600e211ebac737cf7f0020a107000000000000000000000000000001"
				],
				[
					"0x327473e57e0ba6cc39e52adc79597a772c5de123c468aef7f3ac2ab3a76f87cef5aaaa712a621ddb997a94c5b9f0b55f55319b9fdca70ecc03e338c9b8f710176f59a0f4dd080a82",
					"0x93dda1265d8186c0a711e787aae066b72a8d3b4ea57b341b7d71c06220a006d8929b1ca32c496c4a85e9a90d2b5cd4b37f236597bca732f03b561f2855e8df0b929b1ca32c496c4a85e9a90d2b5cd4b37f236597bca732f03b561f2855e8df0b402c4206000000000000000000000000000001"
				]
			]
			```
		- `Reasons` Old Data
			```json
			[
				[
					"0x327473e57e0ba6cc39e52adc79597a77d834d1db4313872258a93b9fc45d488b0095e6fbc1a9e1c46b467c20f936afd09eaddec481acf7a7af9fb0568f1ede3d",
					"0x6501436f6e74726163742042756720666f756e643a204c616e64206f776e65722063616e207265706c616365206f746865727327206472696c6c73202868747470733a2f2f7472656c6c6f2e636f6d2f632f5865366d4d494f5529"
				],
				[
					"0x327473e57e0ba6cc39e52adc79597a77d834d1db4313872258a93b9fc45d488b2b1cf536e9043fc21205e9b698bacde8c286e2e14c4dfb92fe4bcd0de74ecaba",
					"0xfd01392c3939392072696e6720207469702e203320636f6c6c61626f726174696f6e73207769746820746f70203320796f757475626572732074686174206172652077696c6c696e6720746f206d616b6520766964656f7320616e64207368696c6c2064617277696e6961206265666f72652070617261636861696e20736c6f74"
				],
				[
					"0x327473e57e0ba6cc39e52adc79597a77d834d1db4313872258a93b9fc45d488b93dda1265d8186c0a711e787aae066b72a8d3b4ea57b341b7d71c06220a006d8",
					"0xa1014150492042756720666f756e643a20536f6c642061706f73746c65207472616e73616374696f6e73206e6f742073686f77696e67207570206f6e20486973746f7279206d6f64616c202868747470733a2f2f7472656c6c6f2e636f6d2f632f4633465849446a5a29"
				],
				[
					"0x327473e57e0ba6cc39e52adc79597a77d834d1db4313872258a93b9fc45d488bc47094c8866909e63651aea39a57844eb4262d293796ab651bd5d64f7201d7cd",
					"0xe0666f72206a756c792032377468206d65657475702068747470733a2f2f6c696e6b2e6d656469756d2e636f6d2f6d4d73634837796c4c6962"
				],
				[
					"0x327473e57e0ba6cc39e52adc79597a77d834d1db4313872258a93b9fc45d488bee6d208478391c285c1759f7ad54f6ade35f3835d059053f7527a3924b65411c",
					"0xfc466f7220207468652039746820204a756c7920206d65657475707b68747470733a2f2f6c696e6b2e6d656469756d2e636f6d2f314f726c496f53307369627d"
				]
			]
			```
	- **Crab**
		- `ProposalCount` Old Data
			```json
			[
				"0x327473e57e0ba6cc39e52adc79597a776254e9d55588784fa2a62b726696e2b1",
				"0x02000000"
			]
			```
		- `Proposals`
			***Empty***
		- `Approvals` Old Data
			***Empty***
		- `Tips` Old Data
			```json
			[
				[
					"0x327473e57e0ba6cc39e52adc79597a772c5de123c468aef7f3ac2ab3a76f87ce741b65aababcf893747a971ef1a87fb37098d3980e2eeee493b53b71d12fcb1c93c9f4a41cb78317",
					"0xcd8fc3e58dac3b68789bf8a46a724cbb2242daf15ea70f3f34f9e9ba4375550628318c0ecba98d69594de405b8afbb7b80358d4570f164bf9ebca95750362d4228318c0ecba98d69594de405b8afbb7b80358d4570f164bf9ebca95750362d42005a6b3f000000000000000000000000000c762fd1b5bd1db180f27e9ee1c7a4196d85b86c6f9f55048e7733d4816bc2ff5700e87648170000000000000000000000b4f7f03bebc56ebe96bc52ea5ed3159d45a0ce3a8d7f082983c33ef13327474700e87648170000000000000000000000cec66c9be4c0500c5da90cb87d1d0af693da4d5fc9bf0e43ec6195cf7ee4d65000e8764817000000000000000000000001"
				],
				[
					"0x327473e57e0ba6cc39e52adc79597a772c5de123c468aef7f3ac2ab3a76f87cea0dc78da4d91d59864aa5be846eda9424406d6ca916ec2d677b37faa770af65692f895b91a339ee7",
					"0xb9a62fbf857922272fcc82c356377359cada7bb8139f3ddca39a122c4adbc14140d83f0811caa950299dd15ffa41d9cee5e76cdde8b821bb3dd603d3be5a5f6b40d83f0811caa950299dd15ffa41d9cee5e76cdde8b821bb3dd603d3be5a5f6bc0448d40000000000000000000000000000c762fd1b5bd1db180f27e9ee1c7a4196d85b86c6f9f55048e7733d4816bc2ff570010a5d4e80000000000000000000000b4f7f03bebc56ebe96bc52ea5ed3159d45a0ce3a8d7f082983c33ef1332747470010a5d4e80000000000000000000000cec66c9be4c0500c5da90cb87d1d0af693da4d5fc9bf0e43ec6195cf7ee4d6500010a5d4e8000000000000000000000001"
				]
			]
			```
		- `Reasons` Old Data
			```json
			[
				[
					"0x327473e57e0ba6cc39e52adc79597a77d834d1db4313872258a93b9fc45d488bb9a62fbf857922272fcc82c356377359cada7bb8139f3ddca39a122c4adbc141",
					"0x4d0143726561746564206120766964656f206578706c61696e696e6720686f7720746f206275792f7374616b652052494e472f4352494e472068747470733a2f2f796f7574752e62652f3355764a75743550744e34"
				],
				[
					"0x327473e57e0ba6cc39e52adc79597a77d834d1db4313872258a93b9fc45d488bcd8fc3e58dac3b68789bf8a46a724cbb2242daf15ea70f3f34f9e9ba43755506",
					"0x01015b323032312d30345d20436f6d6d756e6974792061637469766974792066756e64696e673a206163746976697479207265776172642d35303030204352494e47"
				]
			]
			```

### Darwinia Companion
https://github.com/darwinia-network/darwinia/pull/710

### Related 
https://github.com/paritytech/substrate/pull/7536
https://github.com/paritytech/polkadot/pull/2025

### CODE REVIEW

**Comment by AurevoirXavier** [comment]
Interesting. These pallets share the same prefix.

File: bin/node/cli/src/chain_spec.rs [hunk 1]
```diff
@@ -770,6 +774,8 @@ pub fn pangolin_local_testnet_config() -> PangolinChainSpec {
 				phantom: PhantomData::<pangolin_runtime::TechnicalMembershipInstance>,
 				members: collective_members.clone(),
 			},
+			pallet_treasury: Default::default(),
```
Comment (AsceticBear) [inline bin/node/cli/src/chain_spec.rs:777]: Why not `pallet_treasury_Instance1`？ like `pallet_collective_Instance1`, `pallet_collective_Instance2`.

Comment (AurevoirXavier) [reply bin/node/cli/src/chain_spec.rs:777]: https://github.com/darwinia-network/darwinia-common/pull/759/files#diff-b33acff65d93ddb65c199e2c906bbc3c862ff6f287dd364a34c4c6e6bfcf46f1R17

File: frame/staking/src/lib.rs [hunk 2]
```diff
@@ -463,7 +459,7 @@ pub trait Config: frame_system::Config + SendTransactionTypes<Call<Self>> {
 	type KtonCurrency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
 	// FIXME: Ugly hack due to https://github.com/rust-lang/rust/issues/31844#issuecomment-557918823
```
Comment (AsceticBear) [inline frame/staking/src/lib.rs:460]: This `FIXME` can be deleted?

File: bin/node/runtime/pangolin/src/pallets/treasury.rs [hunk 3]
```diff
@@ -1,57 +1,50 @@
...
-	type BountyValueMinimum = BountyValueMinimum;
-	type RingBurnDestination = ();
-	type KtonBurnDestination = ();
+	type BurnDestination = Society;
+	type WeightInfo = SubstrateWeight<Runtime>;
+	type SpendFunds = Bounties;
 	type MaxApprovals = MaxApprovals;
+}
+impl Config<KtonTreasuryInstance> for Runtime {
+	type PalletId = TreasuryPalletId;
```
Comment (AsceticBear) [inline bin/node/runtime/pangolin/src/pallets/treasury.rs:36]: For different instances, do we need to choose a different pallet id?

Comment (AurevoirXavier) [reply bin/node/runtime/pangolin/src/pallets/treasury.rs:36]: No need to. A signle account can handle 2 currencies.

File: bin/node/runtime/pangolin/src/lib.rs [hunk 4]
```diff
@@ -899,28 +902,92 @@ impl dvm_rpc_runtime_api::ConvertTransaction<OpaqueExtrinsic> for TransactionCon
...
+				bond: old_proposal.kton_bond,
+			};
+
+			migration::put_storage_value(b"Instance2Treasury", b"Proposals", &hash, new_proposal);
+		}
+	}
+	migration::remove_storage_prefix(OLD_PREFIX, b"Proposals", &[]);
+	log::info!("`Proposals` Migrated");
+
+	migration::move_storage_from_pallet(b"Tips", OLD_PREFIX, NEW_PREFIX);
```
Comment (AsceticBear) [inline bin/node/runtime/pangolin/src/lib.rs:968]: `const NEW_PREFIX: &[u8] = b"Treasury";`  This line move `Tips` to Treasury pallet storage. Is that expected?

I think we should move the storage item to `pallet_tips`.

File: bin/node/runtime/pangolin/src/lib.rs [hunk 5]
```diff
@@ -899,28 +902,92 @@ impl dvm_rpc_runtime_api::ConvertTransaction<OpaqueExtrinsic> for TransactionCon
...
+	log::info!("`Tips` Migrated");
+	migration::move_storage_from_pallet(b"BountyCount", OLD_PREFIX, NEW_PREFIX);
```
Comment (AsceticBear) [inline bin/node/runtime/pangolin/src/lib.rs:970]: And the following bounty related storage.

File: bin/node/runtime/pangolin/src/lib.rs (hunk 4 above)
Comment (AurevoirXavier) [reply bin/node/runtime/pangolin/src/lib.rs:968]: They belong to `Treasury`. 

And I follow the upstream(parity)'s changes.

**Comment by AurevoirXavier** [comment]
@HackFisher Please review.

**Review by HackFisher — APPROVED** [review]

File: bin/node/runtime/pangolin/src/lib.rs [hunk 6]
```diff
@@ -294,7 +294,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	spec_version: 2600,
```
Comment (HackFisher) [inline bin/node/runtime/pangolin/src/lib.rs:294]: Might need bump spec if necessary

**Review by HackFisher — APPROVED** [review]

File: bin/node/runtime/pangolin/src/lib.rs (hunk 6 above)
Comment (AurevoirXavier) [reply bin/node/runtime/pangolin/src/lib.rs:294]: Will be done in release PR.

File: bin/node/runtime/pangolin/src/lib.rs (hunk 4 above)
Comment (AsceticBear) [reply bin/node/runtime/pangolin/src/lib.rs:968]: The migration logic is a little confusing to me. But other changes all LGTM.

Comment (HackFisher) [reply bin/node/runtime/pangolin/src/lib.rs:968]: Also confuse to me, need to re-check, are there tip storage in Treasuary pallet?

Comment (AsceticBear) [reply bin/node/runtime/pangolin/src/lib.rs:968]: https://github.com/darwinia-network/substrate/blob/main/frame/treasury/src/lib.rs#L178 No Tips

Comment (AurevoirXavier) [reply bin/node/runtime/pangolin/src/lib.rs:968]: https://github.com/darwinia-network/substrate/blob/main/frame/tips/src/lib.rs#L139

Comment (AurevoirXavier) [reply bin/node/runtime/pangolin/src/lib.rs:968]:
```rust
as Treasury
```

File: bin/node/runtime/pangolin/src/lib.rs (hunk 5 above)
Comment (HackFisher) [reply bin/node/runtime/pangolin/src/lib.rs:970]: same question

File: bin/node/runtime/pangolin/src/lib.rs (hunk 4 above)
Comment (AurevoirXavier) [reply bin/node/runtime/pangolin/src/lib.rs:968]: And ci-bot will show you what it actually is all the data grab from real metadata.

File: bin/node/runtime/pangolin/src/lib.rs (hunk 5 above)
Comment (AurevoirXavier) [reply bin/node/runtime/pangolin/src/lib.rs:970]: Same https://github.com/darwinia-network/substrate/blob/main/frame/bounties/src/lib.rs#L193

File: bin/node/runtime/pangolin/src/lib.rs (hunk 4 above)
Comment (AsceticBear) [reply bin/node/runtime/pangolin/src/lib.rs:968]: Interesting!!! This is the first time I saw this. Let me check again.

**Review by AsceticBear — APPROVED** [review]

File: bin/node/runtime/pangolin/src/lib.rs (hunk 4 above)
Comment (AurevoirXavier) [reply bin/node/runtime/pangolin/src/lib.rs:968]: It's always good to examine roots and inquire at the base.

**Pull request merged**